### PR TITLE
Feature/model python auto add

### PR DESCRIPTION
**Pull request by vbkaisetsu**
This branch contains two changes for `Model` class in Python:

* Removes `params` and `submodels` properties, and adds `__getitem__` special method that can return both parameters and submodels.
* Adds `add_all_parameters` and `add_all_submodels` methods that automatically register parameters and submodels using attribute information of a Model instance.

### CODE REVIEW

File: python-primitiv/primitiv/_model.pyx [hunk 1]
```diff
@@ -79,15 +39,35 @@ cdef class _Model:
...
+    def add_all_submodels(self):
+        for k, v in self.__dict__.items():
+            if isinstance(v, _Model) and v not in self.added_submodels:
+                self.add_submodel(k, v)
 
-    # NOTE(vbkaisetsu):
-    # get_submodel is replaced with `submodels` variable.
+    def __getitem__(self, key):
+        cdef vector[string] names
+        if isinstance(key, str):
```
Comment (odashi) [inline python-primitiv/primitiv/_model.pyx]: Control flow looks confusing and may generate an incorrect error message.
```
if key has str-type: add key to names
elif key has tuple-type: add elements in key to names
else: throw error with message "incorrect type"

try: get parameter with name `names`
catch:
  try: get submodel with name `names`
  catch: throw error with message "`names` is not a name of neither parameter nor submodel."
```

File: python-primitiv/tests/model.py [hunk 2]
```diff
@@ -89,13 +87,13 @@ def test_model_parameter_deep(self):
         model2.add_submodel("m3", model3)
         param = Parameter()
         model3.add_parameter("p", param)
-        self.assertIs(model1.params["m2", "m3", "p"], param)
+        self.assertIs(model1["m2", "m3", "p"], param)
```
Comment (odashi) [inline python-primitiv/tests/model.py]: The following expression should also be tested: `model1["m2"]["m3"]["p"]`

File: python-primitiv/tests/model.py [hunk 3]
```diff
@@ -105,26 +103,18 @@ def test_model_submodel_deep(self):
         model2.add_submodel("m3", model3)
         model4 = Model()
         model3.add_submodel("m4", model4)
-        self.assertIs(model1.submodels["m2", "m3", "m4"], model4)
+        self.assertIs(model1["m2", "m3", "m4"], model4)
```
Comment (odashi) [inline python-primitiv/tests/model.py]: ditto

File: python-primitiv/tests/model.py [hunk 4]
```diff
@@ -79,7 +77,7 @@ def test_model_parameter(self):
         model = Model()
         param = Parameter()
         model.add_parameter("p", param)
-        self.assertIs(model.params["p"], param)
+        self.assertIs(model["p"], param)
```
Comment (odashi) [inline python-primitiv/tests/model.py]: The following expression should also be tested: `model[("p",)]`

File: python-primitiv/tests/model.py [hunk 5]
```diff
@@ -144,3 +134,31 @@ def test_model_get_all_parameters(self):
         self.assertIs(params[("p2",)], parentmodel.p2)
         self.assertIs(params[("sub", "sp1")], parentmodel.sub.sp1)
```
Comment (odashi) [inline python-primitiv/tests/model.py]: Redundant parenthesis?

Comment (vbkaisetsu) [reply python-primitiv/tests/model.py]: Of course it works without parentheses if tuples contain two or more items.
However `params` is just a dictionary, so it does not work with a single string.

For consistency, it is better to write parentheses.

**Comment by vbkaisetsu** [comment]
Fixed some parts.

File: python-primitiv/primitiv/_model.pyx [hunk 6]
```diff
@@ -79,15 +39,41 @@ cdef class _Model:
...
+                names.push_back(pystr_to_cppstr(name))
+        else:
+            raise TypeError("Argument 'key' has incorrect type (str or tuple)")
+        try:
+            return _Parameter.get_wrapper(&self.wrapped.get_parameter(names))
+        except:
+            try:
+                return _Model.get_wrapper(&self.wrapped.get_submodel(names))
+            except:
+                # NOTE(vbkaisetsu): DO NOT throw an exception here because
```
Comment (odashi) [inline python-primitiv/primitiv/_model.pyx]: If `get_parameter` function did not find a parameter with name `names`, the message from `get_parameter` would be killed and users may get an incorrect error message:  "**Model** with name `names` does not exist."

File: python-primitiv/tests/model.py (hunk 5 above)
Comment (odashi) [reply python-primitiv/tests/model.py]: I meant that this comment points only lines that use 2-tuples, but your argument is also appropriate.

File: python-primitiv/primitiv/_model.pyx (hunk 6 above)
Comment (vbkaisetsu) [reply python-primitiv/primitiv/_model.pyx]: No. Exceptions thrown by `get_submodel()` also be killed.
I fixed a comment and a typo, and added checking the message in the test case.
Does it work good?
```python
from primitiv import *
m = Model()
p = Parameter()
m.add_parameter("p", p)
sm = Model()
m.add_submodel("sm", sm)
m["p"]
# => <primitiv._parameter._Parameter at 0x7fdf04068870>
m["sm"]
# => <primitiv._model._Model at 0x7fdf040619f8>
m["xyz"]
# ---------------------------------------------------------------------------
# TypeError                                 Traceback (most recent call last)
# <ipython-input-9-86188247dc8a> in <module>()
# ----> 1 m["xyz"]
# 
# primitiv/_model.pyx in primitiv._model._Model.__getitem__()
# 
# TypeError: 'name' is not a name of neither parameter nor submodel

```

**Pull request merged**